File: 10283/python/10283.py
#!/usr/bin/python
# 10283 -- ALL IZZ WELL -- Problem code: ALLIZWEL

import sys

def search (matrix, used, cell, position):
	if position is 9:
		return True

	used.add(cell)
	strg = 'ALLIZZWELL'
	#print ("{} --> {}".format(cell,strg[position]))
	x,y = cell
	if (x-1 >= 0):
		if (y-1 >= 0) and matrix[y-1][x-1] == strg[position+1] and not (x-1,y-1) in used:
			if search(matrix, used, (x-1, y-1), position+1): return True
		if matrix[y][x-1] == strg[position+1] and not (x-1,y) in used:
			if search(matrix ,used, (x-1, y) , position+1): return True
		if (y+1 < len(matrix)) and matrix[y+1][x-1] == strg[position+1] and not (x-1,y+1) in used:
			if search(matrix, used, (x-1, y+1), position+1): return True
	if (y-1 >= 0) and matrix[y-1][x] == strg[position+1] and not (x, y-1) in used:
		if search(matrix, used, (x, y-1), position+1): return True
	if (y+1 < len(matrix)) and matrix[y+1][x] == strg[position+1] and not (x,y+1) in used:
		if search(matrix, used, (x, y+1), position+1): return True
	if (x+1 < len(matrix[0])):
		if (y-1 >= 0) and matrix[y-1][x+1] == strg[position+1] and not (x+1,y-1) in used:
			if search(matrix, used, (x+1, y-1), position+1): return True
		if matrix[y][x+1] == strg[position+1] and not (x+1,y) in used:
			if search(matrix ,used, (x+1, y) , position+1): return True
		if (y+1 < len(matrix)) and matrix[y+1][x+1] == strg[position+1] and not (x+1,y+1) in used:
			if search(matrix, used, (x+1, y+1), position+1): return True
	used.remove(cell)
	#print ('{} failed'.format(cell))
	return False


def IzzAllWell (matrix, m, n):
	for x in range(n):
		for y in range(m):
			used = set()
			if matrix[y][x] == 'A' and search(matrix, used, (x,y), 0):
				return True
	return False


def main():
    numberOfCases = int(sys.stdin.readline())
    for i in range(numberOfCases):
	    line = sys.stdin.readline().strip()
	    [M,N] = [int(num) for num in line.split()]
	    lines = []
	    for j in range(M):
		    lines.append(sys.stdin.readline().strip())
	    if IzzAllWell (lines, M, N):
		    print ('YES')
	    else:
		    print ('NO')
	    sys.stdin.readline()
        

if __name__ == '__main__':
    main()
                
    

File: 10283/C/10283.c
#include <stdio.h>

char* strg = "ALLIZZWELL";
int nRows;
int nCols;
int IzzAllWell (char** matrix, char** used, int nRows, int nCols);
int search (char** matrix, char** used, int x, int y, int position);

int main(void) {
  int nCases;
  scanf("%d\n", &nCases);
  int counter;
  for (counter = 0; counter < nCases; counter++) {
    scanf("%d %d\n", &nRows, &nCols);
    char strs [nRows][nCols+1];
    char* matrix [nRows];
    char used_ [nRows][nCols];
    char* used [nRows];
    int rowCounter, colCounter;
    for (rowCounter = 0; rowCounter < nRows; rowCounter++) {
      scanf("%s\n", strs[rowCounter]);
      matrix[rowCounter] = &strs[rowCounter][0];

      for (colCounter = 0; colCounter < nCols; colCounter++)
	used_[rowCounter][colCounter] = 0;
      used[rowCounter] = &used_[rowCounter][0];
    }

    if (IzzAllWell(matrix, used, nRows, nCols))
      printf("YES\n");
    else
      printf("NO\n");
  }
  return 0;
}

int IzzAllWell (char** matrix, char ** used, int nRows, int nCols) {
  int rowCounter, colCounter;
  for (rowCounter = 0; rowCounter < nRows; rowCounter++)
    for (colCounter = 0; colCounter < nCols; colCounter++)
      if (matrix[rowCounter][colCounter] == 'A' && search (matrix, used, colCounter, rowCounter, 0))
	return 1;

  return 0;
}

int search (char** matrix, char** used, int x, int y, int position) {
  if (position == 9)
    return 1;

  used[y][x] = 1;
  
  if (x-1 >= 0){
    if (y-1 >= 0 && matrix[y-1][x-1] == strg[position+1] && !used[y-1][x-1])
      if (search(matrix, used, x-1, y-1, position+1))
	return 1;
    if (matrix[y][x-1] == strg[position+1] && !used[y][x-1])
      if (search(matrix ,used, x-1, y , position+1))
	return 1;
    if ((y+1 < nRows) && matrix[y+1][x-1] == strg[position+1] && !used[y+1][x-1])
      if (search(matrix, used, x-1, y+1, position+1))
	return 1;
  }

  if ((y-1 >= 0) &&  matrix[y-1][x] == strg[position+1] && !used[y-1][x])
    if (search(matrix, used, x, y-1, position+1))
      return 1;

  if ((y+1 < nRows) && matrix[y+1][x] == strg[position+1] && !used[y+1][x])
    if (search(matrix, used, x, y+1, position+1))
      return 1;

  if (x+1 < nCols) {
    if ((y-1 >= 0) && matrix[y-1][x+1] == strg[position+1] && !used[y-1][x+1])
      if (search(matrix, used, x+1, y-1, position+1)) 
	return 1;
    if (matrix[y][x+1] == strg[position+1] && !used[y][x+1])
      if (search(matrix ,used,  x+1, y  , position+1))
	return 1;
    if ((y+1 < nRows)  && matrix[y+1][x+1] == strg[position+1] && !used[y+1][x+1])
      if (search(matrix, used, x+1, y+1, position+1)) 
	return 1;
  }

  used[y][x] = 0;

  return 0;
}
